# Домашнє завдання до теми "Основи роботи з даними у Pandas"

In [1]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


В цьому домашньому завданні продовжуємо працювати з набором даних `supermarket_sales.csv`.

0. Імпортуйте бібліотеку pandas та зчитайте дані у змінну `df` типу `pandas.DataFrame`.

In [11]:
import pandas as pd
df = pd.read_csv('./drive/MyDrive/supermarket_sales.csv')

1. За допомогою бібліотеки Pandas виведіть перші 4 і останні 4 записи у наборі даних (це дві окремі команди). Треба саме 4! Не 5.

In [12]:
df.head(4)

Invoice ID Branch       City Customer type  Gender  \
0  750-67-8428      A     Yangon        Member  Female   
1  226-31-3081      C  Naypyitaw        Normal  Female   
2  631-41-3108      A     Yangon        Normal    Male   
3  123-19-1176      A     Yangon        Member    Male   

             Product line  Unit price  Quantity   Tax 5%     Total       Date  \
0       Health and beauty       74.69         7  26.1415  548.9715   1/5/2019   
1  Electronic accessories       15.28         5   3.8200   80.2200   3/8/2019   
2      Home and lifestyle       46.33         7  16.2155  340.5255   3/3/2019   
3       Health and beauty       58.22         8  23.2880  489.0480  1/27/2019   

    Time      Payment    cogs  gross margin percentage  gross income  Rating  
0  13:08      Ewallet  522.83                 4.761905       26.1415     9.1  
1  10:29         Cash   76.40                 4.761905        3.8200     9.6  
2  13:23  Credit card  324.31                 4.761905       16.2155     7.4  
3  20:33      Ewallet  465.76                 4.761905       23.2880     8.4

In [13]:
df.tail(4)

Invoice ID Branch      City Customer type  Gender         Product line  \
996  303-96-2227      B  Mandalay        Normal  Female   Home and lifestyle   
997  727-02-1313      A    Yangon        Member    Male   Food and beverages   
998  347-56-2442      A    Yangon        Normal    Male   Home and lifestyle   
999  849-09-3807      A    Yangon        Member  Female  Fashion accessories   

     Unit price  Quantity  Tax 5%     Total       Date   Time  Payment  \
996       97.38        10  48.690  1022.490   3/2/2019  17:16  Ewallet   
997       31.84         1   1.592    33.432   2/9/2019  13:22     Cash   
998       65.82         1   3.291    69.111  2/22/2019  15:33     Cash   
999       88.34         7  30.919   649.299  2/18/2019  13:28     Cash   

       cogs  gross margin percentage  gross income  Rating  
996  973.80                 4.761905        48.690     4.4  
997   31.84                 4.761905         1.592     7.7  
998   65.82                 4.761905         3.291     4.1  
999  618.38                 4.761905        30.919     6.6

Очікуваний результат:

![](https://drive.google.com/uc?export=view&id=1F5XlzBOOTMsedpsYqz6NnXjIPTDe6jpx)

2. Виведіть рядки даних з індексами з 615 по 619 включно.

In [17]:
df.iloc[615:620]

Invoice ID Branch       City Customer type  Gender  \
615  509-29-3912      A     Yangon        Member  Female   
616  327-40-9673      B   Mandalay        Member    Male   
617  840-19-2096      C  Naypyitaw        Member    Male   
618  828-46-6863      A     Yangon        Member    Male   
619  641-96-3695      C  Naypyitaw        Member  Female   

               Product line  Unit price  Quantity   Tax 5%     Total  \
615       Sports and travel       38.72         9  17.4240  365.9040   
616       Sports and travel       72.60         6  21.7800  457.3800   
617  Electronic accessories       87.91         5  21.9775  461.5275   
618      Food and beverages       98.53         6  29.5590  620.7390   
619     Fashion accessories       43.46         6  13.0380  273.7980   

          Date   Time      Payment    cogs  gross margin percentage  \
615  3/20/2019  12:24      Ewallet  348.48                 4.761905   
616  1/13/2019  19:51         Cash  435.60                 4.761905   
617  3/14/2019  18:10      Ewallet  439.55                 4.761905   
618  1/23/2019  11:22  Credit card  591.18                 4.761905   
619   2/7/2019  17:55      Ewallet  260.76                 4.761905   

     gross income  Rating  
615       17.4240     4.2  
616       21.7800     6.9  
617       21.9775     4.4  
618       29.5590     4.0  
619       13.0380     8.5

3. Виведіть назви колонок у наборі даних разом з їхніми типами. Це можна зробити кількома способами, важливо аби в результаті були назви колонок і відповідні їм типи.

In [18]:
df.dtypes

Invoice ID                  object
Branch                      object
City                        object
Customer type               object
Gender                      object
Product line                object
Unit price                 float64
Quantity                     int64
Tax 5%                     float64
Total                      float64
Date                        object
Time                        object
Payment                     object
cogs                       float64
gross margin percentage    float64
gross income               float64
Rating                     float64
dtype: object

Очікуваний результат (один з можливих форматів):

![](https://drive.google.com/uc?export=view&id=1ASh6-S-_f3X9fTQAPIvnHMFZamTBjNkL)

4. Користуючись методом `DataFrame.describe()` виявіть 1, 25, 50, 75 і 99 персентилі колонки `gross income`. Знайдіть, яке значення відповідає 75отму персентилю цієї колонки і що це означає? Як би Ви його пояснили простими словами колегам? Відповідь напишіть текстом в окремій клітинці типу Markdown.

In [19]:
df.describe()

Unit price     Quantity       Tax 5%        Total        cogs  \
count  1000.000000  1000.000000  1000.000000  1000.000000  1000.00000   
mean     55.672130     5.510000    15.379369   322.966749   307.58738   
std      26.494628     2.923431    11.708825   245.885335   234.17651   
min      10.080000     1.000000     0.508500    10.678500    10.17000   
25%      32.875000     3.000000     5.924875   124.422375   118.49750   
50%      55.230000     5.000000    12.088000   253.848000   241.76000   
75%      77.935000     8.000000    22.445250   471.350250   448.90500   
max      99.960000    10.000000    49.650000  1042.650000   993.00000   

       gross margin percentage  gross income      Rating  
count              1000.000000   1000.000000  1000.00000  
mean                  4.761905     15.379369     6.97270  
std                   0.000000     11.708825     1.71858  
min                   4.761905      0.508500     4.00000  
25%                   4.761905      5.924875     5.50000  
50%                   4.761905     12.088000     7.00000  
75%                   4.761905     22.445250     8.50000  
max                   4.761905     49.650000    10.00000

Відповідь: для 75% наявних даних такий показник як gross income становить 22.44 та менше.

5. Виведіть кількість інвойсів за кожним унікальним містом в наборі даних.

  Підказка: інвойс - це один рядок в цьому наборі даних, ми можемо цим скористатись аби швидко отримати відповідь.

In [29]:
city_invoice_counts = df['City'].value_counts()
city_invoice_counts

Yangon       340
Mandalay     332
Naypyitaw    328
Name: City, dtype: int64

6. Виведіть, який відсоток інвойсів припадає на кожне унікальне місто.

In [32]:
city_invoice_percentages = df['City'].value_counts(normalize=True) * 100
city_invoice_percentages

Yangon       34.0
Mandalay     33.2
Naypyitaw    32.8
Name: City, dtype: float64

7.1. Використовуючи фільтрацію даних, знайдіть середнє значення `gross income` для міст `Yangon` та `Mandalay`. В якому місті воно більше?

7.2 Сформулюйте 3 гіпотези, чому в одному місті gross income вище, як можна дізнатись чи правдива ця гіпотеза з використанням даних (наявних і додаткових), та як висновки з цих гіпотез можна використати для прийняття рішень як підвищити прибуток в мережі супермаркетів. Відповідь напишіть текстом в окремій клітинці типу Markdown.

Це завдання - на продуктове мислення. Якщо не знаєте, які гіпотези сформулювати - напишіть ті, які здаються вам зовсім примітивними. Важливо навчитись висловлювати свої гіпотези щодо даних, і з часом та досвідом вони будуть ставати щоразу глибше і точніше. Весь аналіз даних будується на тому, що ми перевіряємо певні гіпотези та робимо висновки і рекоемндації на основі своїх досліджень.

In [139]:
yangon_mean_gross_income = df[df['City'] == 'Yangon']['gross income'].mean()

mandalay_mean_gross_income = df[df['City'] == 'Mandalay']['gross income'].mean()

f"yangon_mean_gross_income={yangon_mean_gross_income}, mandalay_mean_gross_income = {mandalay_mean_gross_income}"

'yangon_mean_gross_income=14.874001470588235, mandalay_mean_gross_income = 15.232024096385542'

Відповідь 7.1:
Mandalay

Відповідь 7.2:

###Гіпотези, чому різниця в gross income між містами

1.   Різниця в покупцях.

В Mandalay більше покупок, що підтверджується більшою кількістю інвойсів (340 проти 332) та середнім чеком  (319 проти 312). 75 персентилей покупок в Mandalay зроблена на сумму 465 та нижче проти 459 у Yangon.


В обох містах Mandalay більше покупок зробили чоловіки і в обох містах менше покупців типу Member.




In [136]:
df[df['City'] == 'Yangon']['Total'].mean(), df[df['City'] == 'Mandalay']['Total'].mean()


(312.35403088235296, 319.87250602409637)

In [137]:
df[df['City'] == 'Yangon']['Total'].describe(), df[df['City'] == 'Mandalay']['Total'].describe()


(count     340.000000
 mean      312.354031
 std       231.640025
 min        12.694500
 25%       137.489625
 50%       240.828000
 75%       459.340875
 max      1039.290000
 Name: Total, dtype: float64,
 count     332.000000
 mean      319.872506
 std       242.447565
 min        18.637500
 25%       119.936250
 50%       252.882000
 75%       465.585750
 max      1022.490000
 Name: Total, dtype: float64)

In [135]:
filtered_data = df[df['City'] != 'Naypyitaw']

customer_type_percentages = filtered_data.groupby('City')['Customer type'].value_counts(normalize=True)

gender_percentages = filtered_data.groupby('City')['Gender'].value_counts(normalize=True)


customer_type_percentages, gender_percentages

(City      Customer type
 Mandalay  Normal           0.503012
           Member           0.496988
 Yangon    Normal           0.508824
           Member           0.491176
 Name: Customer type, dtype: float64,
 City      Gender
 Mandalay  Male      0.512048
           Female    0.487952
 Yangon    Male      0.526471
           Female    0.473529
 Name: Gender, dtype: float64)

2. Різниця в асортименті

У порівнянні із Yangon, у Mandalay менший gross income приносить Fashion Accessories, Sport and Travel, але більший всі інші: Electronic accessories, Food and beverages, Home and lifestyle, Health and beauty. Найбільший gross income приносить Health and beauty.



In [140]:
filtered_data = df[df['City'] != 'Naypyitaw']
category_profitability_mean = filtered_data.groupby(['Product line', 'City']).agg({'gross income': 'mean'})
category_profitability_mean

gross income
Product line           City                  
Electronic accessories Mandalay     14.763155
                       Yangon       14.537392
Fashion accessories    Mandalay     12.606234
                       Yangon       15.249775
Food and beverages     Mandalay     14.490370
                       Yangon       14.091216
Health and beauty      Mandalay     17.952075
                       Yangon       12.763681
Home and lifestyle     Mandalay     16.713490
                       Yangon       16.422854
Sports and travel      Mandalay     15.351919
                       Yangon       15.635754

3. Сезонність і місцеві події

У Mandalay був більший gross income в лютому 2019 року.

Mandalay	1639
Yangon	1422

Може бути повʼязане із Mahamuni Pagoda Festival. Одна помітна подія, що відбувається неподалік лютого, - це Фестиваль пагоди Махамуні, який зазвичай проводиться наприкінці лютого або на початку березня, залежно від місячного календаря. Цей фестиваль, що центрується навколо храму Будди Махамуні, одного з найсвятіших місць у Мандалаї, включає різноманітні релігійні заходи, традиційні виступи та жвавий місцевий ринок, що виникає навколо території храму.

In [95]:
filtered_data = df[df['City'] != 'Naypyitaw']
df_copy = filtered_data
df_copy['Date'] = pd.to_datetime(df['Date'])
sales_by_date = df_copy.groupby([df_copy['Date'].dt.month, 'City']).agg({'gross income': 'sum'})

sales_by_date

<ipython-input-95-14432700f1b8>:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_copy['Date'] = pd.to_datetime(df['Date'])


gross income
Date City                  
1    Mandalay     1770.2885
     Yangon       1841.9585
2    Mandalay     1639.2510
     Yangon       1421.9105
3    Mandalay     1647.4925
     Yangon       1793.2915

8. Для продовження дослідження різниці у прибутках в магазинах давайте зʼясуємо, чи відрізняється маржа ('gross margin percentage') для різних інвойсів в магазині? Перевірте це програмно на даних виходячи з того функціоналу, який ми вже пройшли у лекціях.

Дайте на основі отриманого результату відповідь: "У всіх всіх інвойсах однакова маржа" або "Маржа відрізняється для різних інвойсів".

У всіх всіх інвойсах однакова маржа

In [101]:
unique_margins = df['gross margin percentage'].nunique()
unique_margins

1

9. Врахуовуючи висновок в попередньому завданні, дослідимо як продаються в цих містах різні категорії товарів. Виведіть відсотковий розпоіділ кількості інвойсів по категоріям товарів `Product line` в містах `Yangon` та `Mandalay`.

In [133]:
filtered_data = df[df['City'].isin(['Yangon', 'Mandalay'])]

filtered_data.groupby(['City'])['Product line'].value_counts(normalize=True)


City      Product line          
Mandalay  Fashion accessories       0.186747
          Sports and travel         0.186747
          Electronic accessories    0.165663
          Health and beauty         0.159639
          Food and beverages        0.150602
          Home and lifestyle        0.150602
Yangon    Home and lifestyle        0.191176
          Electronic accessories    0.176471
          Sports and travel         0.173529
          Food and beverages        0.170588
          Fashion accessories       0.150000
          Health and beauty         0.138235
Name: Product line, dtype: float64